### ISSUE TITLE
Rclone tries to resolve symlinks even though they are stated in an exclude filter

### ISSUE THREAD

**Issue by onjen** [rclone/rclone#6376]
#### Associated forum post
There is a related [forum post](https://forum.rclone.org/t/symbolic-link-stat-error-on-excluded-file-result-in-i-o-error/23539)

#### What is the problem you are having with rclone?
When rclone is called with the `-L` or `--copy-links` flag it tries to resolve symlinks even though they are stated in an exclude filter. I would expect rclone to ignore the symlinks when they are stated in a filter. It looks like rclone first tries to resolve the links and only afterwards applies the filter. This is also stated by @ncw in the related forum post. I didn't find any related issue to the post by Tercio Filho therefore opening this one.

**Steps to reproduce**
1. Creating a broken symlink
```
touch file1
touch file2
ln -s file1 link1
ln -s file2 link2
rm file 1  # to break link1
```
2. Afterwards although specifically excluded by the filter the broken symlink still results in an Error
```
$ rclone ls -L . --exclude link1 -vv
2022/08/11 10:07:11 DEBUG : rclone: Version "v1.53.3-DEV" starting with parameters ["rclone" "ls" "-L" "." "--exclude" "link1" "-vv"]
2022/08/11 10:07:11 DEBUG : Creating backend with remote "."
2022/08/11 10:07:11 NOTICE: Config file "/home/rothe/.config/rclone/rclone.conf" not found - using defaults
2022/08/11 10:07:11 DEBUG : fs cache: renaming cache item "." to be canonical "/tmp/rclone"
2022/08/11 10:07:11 ERROR : link1: Listing error: symlink: stat /tmp/rclone/link1: no such file or directory
        0 file2
        0 link2
2022/08/11 10:07:11 DEBUG : 2 go routines active
2022/08/11 10:07:11 Failed to ls: symlink: stat /tmp/rclone/link1: no such file or directory
```
exit status is 6

#### What is your rclone version (output from `rclone version`)
```
rclone version
rclone v1.53.3-DEV
- os/arch: linux/amd64
- go version: go1.18.1
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Debian 5.18.14-1 (2022-07-23) x86_64

#### Which cloud storage system are you using? (e.g. Google Drive)
None, this is reproducible locally with `rclone ls`

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
I've been meaning to fix this for ages - give this a go

[v1.60.0-beta.6418.388abb576.fix-6376-local-filter](https://beta.rclone.org/branch/fix-6376-local-filter/v1.60.0-beta.6418.388abb576.fix-6376-local-filter/) on branch [fix-6376-local-filter](https://github.com/rclone/rclone/tree/fix-6376-local-filter) (uploaded in 15-30 mins)

**Comment by onjen** [comment]
Yes, working as expected now, thanks for the super quick fix! Should I close the issue, or do you want to keep it open for the pull request?

**Comment by ncw** [comment]
Thanks for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.60

**Comment by hightowe** [comment]
Sorry to bump this old issue, but I am having this exact problem on rclone v1.61.1 and my reading of this issue is that v1.61.1 should be fixed... Am I misunderstanding something? Thx.

**Comment by ncw** [comment]
@hightowe yes it should be fixed. Can you open a [new issue](https://github.com/rclone/rclone/issues/new/choose) with a way to reproduce the issue please.

**Comment by hightowe** [comment]
@ncw - I can reproduce literally exactly what this issue reported. I am happy to create a new issue, but before I do that, here is what I am seeing...

A little script for convenience (essentially c/p'd from this issue):

~~~
hightowe@zion ~/tmp/rclone $ cat issue.sh 
#!/bin/bash

# clean from a prior run
rm -f ./file[0-9] ./link[0-9]

# Make a couple of files and links
touch ./file1
touch ./file2
ln -s ./file1 ./link1
ln -s ./file2 ./link2
rm ./file1  # to break link1

# rclone while excluding the broken link
rclone ls -L ./ --exclude link1 -vv

echo -e "\nrclone exit status: $?\n"

# Report our version
rclone --version
~~~

Running that gives these results:

~~~
hightowe@zion ~/tmp/rclone $ ./issue.sh 
2023/02/24 11:06:55 DEBUG : rclone: Version "v1.61.1" starting with parameters ["rclone" "ls" "-L" "./" "--exclude" "link1" "-vv"]
2023/02/24 11:06:55 DEBUG : Creating backend with remote "./"
2023/02/24 11:06:55 DEBUG : Using config file from "/home/hightowe/.config/rclone/rclone.conf"
2023/02/24 11:06:55 DEBUG : local: detected overridden config - adding "{12rtk}" suffix to name
2023/02/24 11:06:55 DEBUG : fs cache: renaming cache item "./" to be canonical "local{12rtk}:/home/hightowe/tmp/rclone"
2023/02/24 11:06:55 ERROR : link1: Listing error: symlink: stat /home/hightowe/tmp/rclone/link1: no such file or directory
        0 file2
      361 issue.sh
        0 link2
2023/02/24 11:06:55 DEBUG : 2 go routines active
2023/02/24 11:06:55 Failed to ls: symlink: stat /home/hightowe/tmp/rclone/link1: no such file or directory

rclone exit status: 6

rclone v1.61.1
- os/version: linuxmint 19.3 (64 bit)
- os/kernel: 5.4.0-64-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.4
- go/linking: static
- go/tags: none
~~~

PS: I tried rclone source paths of both "./" (as in my script in this comment) and "." (as in this original bug report)  and both give the same "ERROR : link1"

Please let me know if I can help in any other way, and thank you for your work on rclone -- a really great tool!

**Comment by onjen** [comment]
Interesting, it seems like this is somehow not part of the release. So the beta I tested works

```
$ rclone rclone ls -L . --exclude link1 -vv && rclone --version
2023/02/24 21:25:54 DEBUG : rclone: Version "v1.60.0-beta.6418.388abb576.fix-6376-local-filter" starting with parameters ["rclone" "ls" "-L" "." "--exclude" "link1" "-vv"]
2023/02/24 21:25:54 DEBUG : Creating backend with remote "."
2023/02/24 21:25:54 DEBUG : local: detected overridden config - adding "{12rtk}" suffix to name
2023/02/24 21:25:54 NOTICE: Config file "/home/rothe/.config/rclone/rclone.conf" not found - using defaults
2023/02/24 21:25:54 DEBUG : fs cache: renaming cache item "." to be canonical "local{12rtk}:/tmp/rclone"
        0 file2
        0 link2
2023/02/24 21:25:54 DEBUG : 2 go routines active

rclone v1.60.0-beta.6418.388abb576.fix-6376-local-filter
- os/version: debian bookworm/sid (64 bit)
- os/kernel: 6.1.0-3-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19
- go/linking: static
- go/tags: none
```

But release 1.60.1 is not working
```
2023/02/24 21:27:21 DEBUG : rclone: Version "v1.60.1-DEV" starting with parameters ["rclone" "ls" "-L" "." "--exclude" "link1" "-vv"]
2023/02/24 21:27:21 DEBUG : Creating backend with remote "."
2023/02/24 21:27:21 NOTICE: Config file "/home/rothe/.config/rclone/rclone.conf" not found - using defaults
2023/02/24 21:27:21 DEBUG : local: detected overridden config - adding "{12rtk}" suffix to name
2023/02/24 21:27:21 DEBUG : fs cache: renaming cache item "." to be canonical "local{12rtk}:/tmp/rclone"
2023/02/24 21:27:21 ERROR : link1: Listing error: symlink: stat /tmp/rclone/link1: no such file or directory
        0 file2
        0 link2
2023/02/24 21:27:21 DEBUG : 2 go routines active
2023/02/24 21:27:21 Failed to ls: symlink: stat /tmp/rclone/link1: no such file or directory

rclone v1.60.1-DEV
- os/version: debian bookworm/sid (64 bit)
- os/kernel: 6.1.0-3-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.5
- go/linking: dynamic
- go/tags: none
```

**Comment by ncw** [comment]
It looks like 388abb576 (the commit id of the working beta) was merged as 8d1fff9a8237c64f73cc02e273921efde1553487 - these commits are identical so the changes have definitely been merged.

However that commit had a bug in it and needed to be fixed in a61d219bcd3ffa3ee0411e7d3759c0ef5e22e495 which is probably where the problem happened.

A bit of testing later, and I can confirm that a61d219bcd3ffa3ee0411e7d3759c0ef5e22e495 is definitely the problem.

I've had a go at fixing this here

[v1.62.0-beta.6741.5e7230150.fix-6376-local-include](https://beta.rclone.org/branch/fix-6376-local-include/v1.62.0-beta.6741.5e7230150.fix-6376-local-include/) on branch [fix-6376-local-include](https://github.com/rclone/rclone/tree/fix-6376-local-include) (uploaded in 15-30 mins)

The fix was simple, but I had to write a load of tests first with dangling symlinks (which I probably should have done before)! The test suite now has the equivalent of the test script of this issue in it.

Please give it a test and report back - thanks!

**Comment by ncw** [comment]
I've merged this fix to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.62
